### DISCUSSION TITLE
podman support?

### DISCUSSION THREAD

**Discussion by igagis** [ChristopherHX/github-act-runner#178] [General]
I'm thinking to install `github-act-runner` on my `riscv64` boards which, AFAIK, has no official `docker` binary build. But, there is `podman` which can easily be installed from debian repo.

So, I'm wondering, will `github-act-runner` transparently use `podman` instead of `docker`? If not, would be great if such support is added. Thanks!

**Comment by ChristopherHX** [comment]
> will `github-act-runner` transparently use `podman` instead of `docker`?

No it won't do that transparently, as nektos/act doesn't really support podman and directly uses docker api via it's go api library.

You need to enable the podman.sock, if you would relocate `/var/run/podman/podman.sock` to `/var/run/docker.sock` then it works transparently. (a symlink itself might behave diffently, due to bind mounts of that location)

Most would work not transparently by setting env `DOCKER_HOST=unix:///var/run/podman/podman.sock` in the service.
I guess your old script has trouble with adding custom envs, but not exactly shure.
The newer svc command has `.env` support to set such stuff, but seems like the sh shim of the deb's hides it.

**Reply by igagis** [reply]
Thanks for your reply!
But I'm not sure I fully understand everything.
What is the approach you recommend to enable `github-act-runer` to use `podman` instead of `docker`?

My understanding is that if `DOCKER_HOST` is set to `podman.sock` in `systemd` service which starts `github-act-runner` then it will use `podman`?

**Reply by ChristopherHX** [reply]
My systems don't have podman installed, since on ubuntu it is always far outdated if installed via apt
> My understanding is that if `DOCKER_HOST` is set to `podman.sock` in `systemd` service which starts `github-act-runner` then it will use `podman`?

yes, that could work.

You might need to run `sudo touch /var/run/docker.sock` in this case if container creation fails.
act does a bind mount of like `-v /var/run/docker.sock:/var/run/docker.sock` that has no public facing setting in github-act-runner. I could probably add such a setting....

**Reply by igagis** [reply]
Tried to set `DOCKER_HOST` to podman in systemd service. Trying to start a job gives this error:
```
Runner Name: lpi4ar0
Runner OSDescription: github-act-runner linux/riscv64
Runner Version: 0.6.10
Initialize translating the job request to nektos/act
Starting nektos/act
🚀  Start image=riscv64/debian:bookworm
  🐳  docker pull image=riscv64/debian:bookworm platform= username= forcePull=true
Error: permission denied while trying to connect to the Docker daemon socket at unix:///var/run/podman/podman.sock: Post "http://%2Fvar%2Frun%2Fpodman%2Fpodman.sock/v1.24/images/create?fromImage=riscv64%2Fdebian&tag=bookworm": dial unix /var/run/podman/podman.sock: connect: permission denied
```
I have added the runner for non-root user.
I'm able to run `podman run hello-world` from the same user without any problems.

The socket is owned by root:
```
ivan@lpi4a:~$ sudo ls -l /var/run/podman/podman.sock
srw-rw---- 1 root root 0 Jun  1 12:49 /var/run/podman/podman.sock
```

There is no any `podman` group I could add my user to (as I do for docker, by adding user to `docker` group).

Any ideas how to cope with this?

**Reply by igagis** [reply]
I did `sudo chmod 777 /var/run/podman && sudo chmod 777 /var/run/podman/podman.sock` and now it works.

The `sudo touch /var/run/docker.sock` is NOT needed.

But, my question still remains. Using `chmod` is a hackish way of doing it. I think there should be safer way to solve the permission problem, but I don't know how.

**Reply by mochja** [reply]
Hey, you can try

```
systemctl --user start podman.socket
export DOCKER_HOST=unix://$XDG_RUNTIME_DIR/podman/podman.sock
```

https://github.com/containers/podman/blob/main/docs/tutorials/socket_activation.md#podman-socket-activation

https://www.server-world.info/en/note?os=Debian_12&p=podman&f=12

Podman although is compatible with docker, it is not docker. I suggest to go through podman documentation for more information.

**Reply by igagis** [reply]
@mochja 
Thanks for suggestion, I think it is a right direction, but unfortunately it didn't work right a way:

```
Runner Name: lpi4ar0
Runner OSDescription: github-act-runner linux/riscv64
Runner Version: 0.6.10
Initialize translating the job request to nektos/act
Starting nektos/act
🚀  Start image=debian:sid
  🐳  docker pull image=debian:sid platform= username= forcePull=true
  🐳  docker create image=debian:sid platform= entrypoint=["tail" "-f" "/dev/null"] cmd=[]
Error: failed to create container: 'Error response from daemon: make cli opts(): making volume mountpoint for volume /var/run/docker.sock: mkdir /var/run/docker.sock: permission denied'
```

No idea what's going on...

Though, it seems that it did connect to `podman.sock`.

**Reply by ChristopherHX** [reply]
like I said act hardcodes a bind mount on `/var/run/docker.sock` and if you specify `DOCKER_HOST` to podman that location won't change it's path


https://github.com/nektos/act/blob/c959fdd58835195916386b44a87632151a1b491a/pkg/runner/runner.go#L47

this opt needs to be changed for podman around here https://github.com/ChristopherHX/github-act-runner/blob/d0e4e4353ae856174e32d997f916dc4ff2801550/actionsdotnetactcompat/act_worker.go#L305

extracting the value out of DOCKER_HOST should work and I can indeed change it

**Reply by igagis** [reply]
Strange thing is that when I did `chmod 777` it just worked. But when I used approach with `systrmctl --user start podman` it started giving this volume mount error

**Reply by ChristopherHX** [reply]
`--user` start rootless podman, rootful podman have more permission and might cause the difference

You might be able to also create a custom docker group, and chown the rootful podman socket so it might be able to mimic docket in this aspect.

Docker is by default always rootful instead of rootless

**Reply by igagis** [reply]
If I do `sudo touch /var/run/docker.sock` then it starts to work.
Now, the question, @ChristopherHX can you make the change to `act` which would allow to avoid creating this dummy `docker.sock`?

I'm planning to make a change to the systemd scripts to allow configuring runner to use podman instead of docker. And in combination with the change in `act` it could provide good user experience.

**Reply by igagis** [reply]
I actually went ahead and made a pull request https://github.com/ChristopherHX/github-act-runner/pull/179 for script side changes

**Reply by igagis** [reply]
All needed changes are already implemented and merged, new runner version released.
I have already tried it with podman, seems working. Yay! Thanks @ChristopherHX and @mochja !
